### ISSUE TITLE
Emulate using WITH in UNION subqueries by wrapping the subquery in a derived table in dialects where this isn't supported

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#10711]
Like #10714, but this approach favours wrapping a UNION in a derived table.

In jOOQ 3.13.5 / 3.14.0, we've fixed generating the correct syntax for `WITH` in `UNION` subqueries. But not all SQL dialects support this in the first place. Some dialects support `WITH` in scalar subqueries and in derived tables, but not in `UNION` subqueries, e.g. this is not possible in HSQLDB (though perfectly valid in H2, PostgreSQL):

```sql
(
  select 1 as "x"
  from (values (1)) t (a)
)
union (
  with "t"("x") as (
    select 1 as "x"
    from (values (1)) t (a)
  )
  select "x"
  from "t"
)
```

There are 2 ways to emulate this with existing infrastructure:

**Pull up the CTE to the top level** (see #10714)

This will work on all dialects, including the ones that cannot nest CTE in derived tables, but produces the risk of naming collisions in case we already have a top-level CTE of that name, or if two union subqueries produce the same top-level CTE names.

**Wrap union subquery in a derived table**

This does not produce any naming collisions but may not work on all RDBMS:

```sql
select 1 as "x"
from (values (1)) t (a)
union 
select * from ( 
  with "t"("x") as (
    select 1 as "x"
    from (values (1)) t (a)
  )
  select "x"
  from "t"
) "t"
```

### Specific changes

- [x] Db2
- [x] HSQLDB
- [x] MariaDB

**Comment by lukaseder** [comment]
This seems to be a risky backport. A lot has changed in `SelectQueryImpl` in jOOQ 3.14. The backport is not worth it, given the relatively esoteric SQL usage.

**Comment by lukaseder** [comment]
Pulling up the CTE will be implemented separately as it is more difficult: #10714
